### ISSUE TITLE
Add additional storage data to monitoring API

### ISSUE THREAD

**Issue by jkuester** [medic/cht-core#9428]
**What feature do you want to improve?**
Currently our [monitoring API](https://github.com/medic/cht-core/blob/66fc2545e63c3df843d952f2bba29dfff0aa87a0/api/src/services/monitoring.js#L331) is returning the following [data about our Couch databases](https://docs.communityhealthtoolkit.org/apps/reference/api/#get-apiv2monitoring):  

```json
  "couchdb": {
    "medic": {
      "name": "medic",
      "update_sequence": 7162,
      "doc_count": 794,
      "doc_del_count": 28,
      "fragmentation": 1.017210715181778
    },
    "sentinel": {
...
    },
    "usersmeta": {
...
    },
    "users": {
...
    }
  },
```

To help better track and understand data usage in the CHT we should add some additional fields.

**Describe the improvement you'd like**

```json
  "couchdb": {
    "medic": {
      "sizes": {
        "file": 108430,
        "active": 882
      },
      "compact_running": false,
      "view_index": {
        "medic": {
          "name": "medic",
          "updater_running": false,
          "compact_running": false,
          "sizes": {
            "file": 8036561,
            "active": 1205784
          }
        },
        "medic-admin": {
          "name": "medic-admin",
          "updater_running": false,
          "compact_running": false,
          "sizes": {
            "file": 8036561,
            "active": 1205784
          }
        },
        "medic-client": {
          "name": "medic-client",
          "updater_running": false,
          "compact_running": false,
          "sizes": {
            "file": 8036561,
            "active": 1205784
          }
        },
        "medic-conflicts": {
          "name": "medic-conflicts",
          "updater_running": false,
          "compact_running": false,
          "sizes": {
            "file": 8036561,
            "active": 1205784
          }
        },
        "medic-scripts": {
          "name": "medic-scripts",
          "updater_running": false,
          "compact_running": false,
          "sizes": {
            "file": 8036561,
            "active": 1205784
          }
        },
        "medic-sms": {
          "name": "medic-sms",
          "updater_running": false,
          "compact_running": false,
          "sizes": {
            "file": 8036561,
            "active": 1205784
          }
        }
      }
    },
    "sentinel": {
      "sizes": {
        "file": 108430,
        "external": 624,
        "active": 882
      },
      "compact_running": false,
      "view_index": {
        "sentinel": {
          "name": "sentinel",
          "updater_running": false,
          "compact_running": false,
          "sizes": {
            "file": 8036561,
            "active": 1205784
          }
        }
      }
    },
    "usersmeta": {
      "sizes": {
        "file": 108430,
        "external": 624,
        "active": 882
      },
      "compact_running": false,
      "view_index": {
        "users-meta": {
          "name": "users-meta",
          "updater_running": false,
          "compact_running": false,
          "sizes": {
            "file": 8036561,
            "active": 1205784
          }
        }
      }
    },
    "users": {
      "sizes": {
        "file": 108430,
        "external": 624,
        "active": 882
      },
      "compact_running": false,
      "view_index": {
        "users": {
          "name": "users",
          "updater_running": false,
          "compact_running": false,
          "sizes": {
            "file": 8036561,
            "active": 1205784
          }
        }
      }
    }
  }
```

- **sizes.active** (*number*) – The size of live data inside the database, in
  bytes.  
- **sizes.external** (*number*) – The uncompressed size of database contents
  in bytes.  
- **sizes.file** (*number*) – The size of the database file on disk in bytes.
  Views indexes are not included in the calculation.

The db-level data can be pulled from the same `_dbs_info` call that we are already making as part of the `/monitoring` flow.  

For the `view_index` data, we will  need to make additional calls to the `{{db-doc}}/_info` endpoints for each of the views to collect this data.

**Describe alternatives you've considered**

If we exclude the view-specific metrics, we would not actually be incurring any additional requests while compiling data. 

**Additional context**
Related to https://github.com/medic/cht-core/issues/9293 and "[Measuring reduced disk usage in TCO](https://docs.google.com/document/d/1X0ztxCdmdGWo_-6PcHkwTTwvbgXQxR_h6W9JZGYEmWU/edit#heading=h.j82fl3pip3ax)" (private google doc)

**Comment by jkuester** [comment]
After some very helpful [clarifications in the Couch Slack channel](https://couchdb.slack.com/archives/C49LEE7NW/p1726147200520889), I am going to amend my original design to remove the `sizes.external` values.  This number represents 

> the size of your data if exported in json (i.e, without the database structure overhead).

This data is not particularly useful in day-to-day maintenance of the CHT and it could cause unnecessary confusion to include this extra number in our metrics.  

Another interesting fact I learned was that this DB-level data (from `_dbs_info`) _does not include any view index data_.  Currently we are calculating the existing db `fragmentation` metric with `fragmentation = sizes.files / sizes.active`.  But this just represents the fragmentation of the database JSON data and not of the view data.  I think this re-emphasises the importance and value of our plan to expose these size metrics _for each view_.

**Comment by jkuester** [comment]
@mrjones-plip @m5r @sugat009 One design question I have for you smart folks:

> Currently we are calculating the existing db fragmentation metric with fragmentation = sizes.files / sizes.active.

Given this fact, do you think we actually need to provide both `sizes.files` and `sizes.active`?  Basically, one of those three fields is technically redundant (since you can calculate it once you know the other two values).  In my reading around Prometheus stuff, it seems like folks are not super concerned about providing redundant metrics (since some use-cases are better addressed with raw data vs calculated).  However, I am more thinking about this from the point of clarity and not confusing to someone trying to understand the data values.  

Basically, instead of 

```json
      "fragmentation": 122,
      "sizes": {
        "file": 108430,
        "active": 882
      },
```

I am considering:

```json
      "fragmentation": 122,
      "size": 882
```

In this case, if you wanted to know the _size of the DB on disk_, you would just do `fragmentation * size`.

**Comment by mrjones-plip** [comment]
Excellent research here  jkuester  - thank you!

My vote is we keep all three.  I imagine they're all going to a bit opaque as to what exactly the number means and that folks will be referencing the docs often to see what is what.  

I don't feel too strongly on this! As it's not a big deal either way,  you should take another ~3 minutes deciding and just go for it :rocket:

**Comment by jkuester** [comment]
@mrjones-plip An interesting counterproposal would be to keep all three values, but update the `fragmentation` metric for each database to also factor in the aggregate fragmentation of all the views included in that database.  With this approach, you would still be able to calculate specific fragmentation values for each individual db/view-index, but you would also have a summary `fragmentation` value that would give a good estimation of the overall data health for that database...

The main downside that I see with this approach is it is kinda sketchy to change the scope for the value being returned in the `fragmentation` field without bumping the version of the `/monitoring` api.  On one hand, the purpose of the field is still the same and the syntax of the value has not changed, but for anyone tracking the `fragmentation` value over time, they may see an abrupt change to the trend with the new approach.  That being said, even with view-index data included, our fragmentation target is still aimed at `1` as it was before and compacting things is still the method for controlling this value.  So, all in all, I am strongly leaning in this direction unless someone has an objection....

**Comment by sugat009** [comment]
> do you think we actually need to provide both sizes.files and sizes.active?

Depends on what's necessary. Are the end users using these? My design thinking is to always give out the raw values and not calculated values like giving the person's DOB and not the age. 

> The main downside that I see with this approach is it is kinda sketchy to change the scope for the value being returned in the fragmentation field without bumping the version of the /monitoring API.

I think we should apply the "open to extension and closed to modification" principle here. We should not be changing anything in the existing API and if we need to then, we should create a newer version of it.

**Comment by jkuester** [comment]
> My design thinking is to always give out the raw values and not calculated values like giving the person's DOB and not the age.

Yeah, this is a good way to put this.  I will go ahead and return `files` and `active` for all the dbs _and also all the view_indexes_ for each db.  

> I think we should apply the "open to extension and closed to modification" principle here.

I agree that this is the _proper approach_ in a case where you are fundamentally changing behavior of an existing api.  However, the more I have considered this case of updating the `fragmentation` value to also include view fragmentation for that DB, the more I would characterize the change as a "bug fix".  The docs for the current `/monitoring` api say:

> The fragmentation of the db, lower is better, “1” is no fragmentation.

Now, the existing api does not currently return any data for _specific view indexes_, in the db. In reality, those view indexes can use more disk space than the actual db data.  I had previously assumed (based on a straightforward/naive reading of the documentation) that the `fragmentation` value was reflective of the total data on disk for the database (db contents + view indexes).  It turns out this is not the case and the current `fragmentation` value is only for the _db contents_.  

To be more clear:
- The simple interpretation of the current api is that the `fragmentation` value represents the _total_ fragmentation of all the db data on disk.
- However, the current `fragmentation` value is actually only reflective of a sub-set of the db's data stored on the disk. (since it does not include the view index data)
- Updating the `fragmentation` value to include view index fragmentation would "fix" the value to actually represent what folks already think it indicates.
- This is why I think we can modify the value returned for `fragmentation` without actually bumping the api version.

**Comment by jkuester** [comment]
FYI, I am considering dropping the `updater_running` and `compact_running` fields from the api.  They never seemed to quite fit with the rest of the data, but it was not until I went to make the Watchdog changes to consume the api that I was able to conceptualize the issue.

The main problem is that when you are _periodically polling_ `/monitoring`, it is not super helpful to be sampling data about an ephemeral task (particularly one that might complete have a runtime that is < the polling interval.  For example, our default polling interval with Watchdog is `1m`.  For compaction on `_users`, it may only take a few seconds to complete and Watchdog would only ever detect it if it happened to poll `/monitoring` at the exact right moment.  Otherwise, in Watchdog it will just look like compaction never ran on `_users`.  To properly collect this kind of data we (or really Couch) should be _pushing_ an entry to some sort of log each time it runs compaction.  (Couch does print to the console when it starts/finishes compaction. Not sure if there is a better place to programmatically access this data.)

That being said, for some troubleshooting tasks, it could be really helpful to know that compaction is running on an index is still being updated.  I am open to the possibility of leaving the data in the `/monitoring` api, but _not_ including it in Watchdog.  My main objection to this approach is that if you are troubleshooting something, what you really want to look at is just Couch's `_active_tasks`.  If we want to support more troubleshooting workflows via the `/monitoring` api, we should consider just exposing the `_active_tasks` somehow and not forcing folks to drill though the whole JSON to figure out which views are still being indexed.....

**Comment by jkuester** [comment]
Okays, so I also ended up changing the `view_index` object for each db to be a `view_indexes` array.  It made more sense semantically and I could not come up with a reason to use an object (other than matching the existing pattern for `couchdb`).  So, the new data structure for the `/monitoring` api should look like:

```json
  "couchdb": {
    "medic": {
      "sizes": {
        "file": 108430,
        "active": 882
      },
      "view_indexes": [
        {
          "name": "medic",
          "sizes": {
            "file": 8036561,
            "active": 1205784
          }
        },
...
      ]
    }
  }
```

This was done in a separate commit from removing the `updater_running` and `compact_running` fields.  So, it should not be hard to add one or both of these back into the branch if we decide to...
